go/vt/vttablet/tabletmanager/rpc_backup.go

```diff
@@ -49,22 +44,28 @@ func (agent *ActionAgent) Backup(ctx context.Context, concurrency int, logger lo
 	if !allowMaster && currentTablet.Type == topodatapb.TabletType_MASTER {
 		return fmt.Errorf("type MASTER cannot take backup. if you really need to do this, rerun the backup command with -allow_master")
 	}
-
-	tablet, err := agent.TopoServer.GetTablet(ctx, agent.TabletAlias)
-	if err != nil {
-		return err
-	}
-	if !allowMaster && tablet.Type == topodatapb.TabletType_MASTER {
-		return fmt.Errorf("type MASTER cannot take backup. if you really need to do this, rerun the backup command with -allow_master")
-	}
-	originalType := tablet.Type
-
 	engine, err := mysqlctl.GetBackupEngine()
 	if err != nil {
 		return vterrors.Wrap(err, "failed to find backup engine")
 	}
 	builtin, _ := engine.(*mysqlctl.BuiltinBackupEngine)
+	// get Tablet info from topo so that it is up to date
+	tablet, err := agent.TopoServer.GetTablet(ctx, agent.TabletAlias)
+	if err != nil {
+		return err
+	}
+	originalType := tablet.Type
 	if builtin != nil {
+		if err := agent.lock(ctx); err != nil {
```

Comment:
This is unrelated to the changes in this PR, but just a thought: instead of special-casing `builtin`, could we add a method like `ShouldDrainForBackup() bool` to the `BackupEngine` interface? I feel like that would make the conditionals here quicker to grasp.

Reply:
another good idea :) we should do that.
@teejae had similar feedback when he reviewed this a while ago.